compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -2408,64 +2444,92 @@ public void visit(BLangTupleDestructure tupleDestructure) {
-            restParamAssignment.expr = arrayExpr;
-
-            // foreach var $foreach$i in tupleTypes.length()...tupleLiteral.length() {
-            //     t[t.length()] = <T> tupleLiteral[$foreach$i];
-            // }
-            BLangLiteral startIndexLiteral = (BLangLiteral) TreeBuilder.createLiteralExpression();
-            startIndexLiteral.value = (long) tupleVarRef.expressions.size();
-            startIndexLiteral.setBType(symTable.intType);
-            BLangInvocation lengthInvocation = createLengthInvocation(pos, tupleExpr);
-            BLangInvocation intRangeInvocation = replaceWithIntRange(pos, startIndexLiteral,
-                    getModifiedIntRangeEndExpr(lengthInvocation));
-
-            BLangForeach foreach = (BLangForeach) TreeBuilder.createForeachNode();
-            foreach.pos = pos;
-            foreach.collection = intRangeInvocation;
-            types.setForeachTypedBindingPatternType(foreach);
-
-            final BLangSimpleVariable foreachVariable = ASTBuilderUtil.createVariable(pos,
-                    "$foreach$i", foreach.varType);
-            foreachVariable.symbol = new BVarSymbol(0, names.fromIdNode(foreachVariable.name),
-                                                    this.env.scope.owner.pkgID, foreachVariable.getBType(),
-                                                    this.env.scope.owner, pos, VIRTUAL);
-            BLangSimpleVarRef foreachVarRef = ASTBuilderUtil.createVariableRef(pos, foreachVariable.symbol);
-            foreach.variableDefinitionNode = ASTBuilderUtil.createVariableDef(pos, foreachVariable);
-            foreach.isDeclaredWithVar = true;
-            BLangBlockStmt foreachBody = ASTBuilderUtil.createBlockStmt(pos);
+        // T[] t = [];
+        BLangSimpleVarRef restParam = (BLangSimpleVarRef) tupleVarRef.restParam;
+        BType restParamType = restParam.getBType();
+        BLangArrayLiteral arrayExpr = createArrayLiteralExprNode();
+        arrayExpr.setBType(restParamType);
+
+        BLangAssignment restParamAssignment = ASTBuilderUtil.createAssignmentStmt(pos, blockStmt);
+        restParamAssignment.varRef = restParam;
+        restParamAssignment.varRef.setBType(restParamType);
+        restParamAssignment.expr = arrayExpr;
+
+        BLangExpression countExpr = ASTBuilderUtil.createLiteral(pos, symTable.intType,
+                (long) tupleVarRef.expressions.size());
+        boolean isIndexBasedAccessExpr = tupleExpr.getKind() == NodeKind.INDEX_BASED_ACCESS_EXPR;
+        if (restParamType.tag == TypeTags.TUPLE) {
+            BTupleType restTupleType = (BTupleType) restParamType;
+            BLangIndexBasedAccess modifiedTupleExpr = ASTBuilderUtil.createIndexAccessExpr(tupleExpr, countExpr);
+            modifiedTupleExpr.setBType(tupleExpr.getBType());
+            tupleExpr = modifiedTupleExpr;
+            isIndexBasedAccessExpr = true;
+            for (int index = 0; index < restTupleType.tupleTypes.size(); index++) {
```

Comment:
Shall we move common logic used here and createRestFieldVarDefStmts() out

Reply:
The logic is a bit different for creating the `var defs` for a rest field in a tuple variable and in a tuple restructure. Therefore it's difficult to move the seemingly common logic given in those methods out.